### DISCUSSION TITLE
how to use adjoint method to optimize surface texture?

### DISCUSSION THREAD

**Discussion by zjuyinge** [NanoComp/meep#2205] [Q&A] [answered]
i want to do surface texture optimization with meep adjoint method,  but matgrid is for density optimization,  how to convert it to optimize the hight of one layer.  surpose mat1 and mat2 are two layers in a 3D model, I want to optimize the height of mat1 h(x, y).

**Comment by smartalecH** [comment, marked as answer]
Hmm doing this out of the box is a bit tricky. 

You could take a look at a [repo](https://github.com/zlin-opt/meep-adjoint-3d) created @zlin-opt which essentially computes the adjoint gradient from scratch using common meep functions specifically for your problem. 

If you did want to use the existing `MaterialGrid` infrastructure, you could do a full 3D optimization with additional connectivity constraints. Although this would require some finesse to get exactly what you want. (And @mochen4 is still debugging some issues with the current implementation).

**Reply by zjuyinge** [reply]
thanks for your kindly reply.
